Add configurable exit grace periods and shutdown handlers

kpayson64 opened the pull request.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_server.py:
```diff
@@ -719,22 +727,26 @@ def _start(state):
+        _stop(state, state.exit_grace, state.exit_handler_grace).wait()
       else:
-        _stop(state, timeout).wait()
+        _stop(state, timeout, True).wait()
```
I don't understand the `True` on this line. Isn't a numeric value or `None` appropriate in that position?

kpayson64 replied:
Oops, good catch.  I changed the API when I was working on this and this is a vestige.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_server.py:
```diff
@@ -719,22 +727,26 @@ def _start(state):
     self._state = _ServerState(
-        completion_queue, server, generic_handlers, thread_pool)
+        completion_queue, server, generic_handlers, thread_pool,
+        exit_grace if exit_grace is not None else _DEFAULT_EXIT_GRACE,
```
Prefer avoiding `not` when you can; flip these branches and drop the `not`.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_server.py:
```diff
@@ -719,22 +727,26 @@ def _start(state):
-        completion_queue, server, generic_handlers, thread_pool)
+        completion_queue, server, generic_handlers, thread_pool,
+        exit_grace if exit_grace is not None else _DEFAULT_EXIT_GRACE,
+        exit_handler_grace if exit_handler_grace is not None else
```
Here too.

kpayson64 commented:
Addressed comments

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/__init__.py:
```diff
@@ -905,6 +905,21 @@ def add_secure_port(self, address, server_credentials):
 
   @abc.abstractmethod
+  def add_shutdown_handler(self, shutdown_handler):
+    """Adds a shutdown handler to be called on server shutdown.
```
Drop the first "handler" in this sentence - the term "shutdown handler" isn't defined until the reader reads the rest of the sentence and infers that being called on server shutdown is what makes a handler a "shutdown handler".

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/__init__.py:
```diff
@@ -905,6 +905,21 @@ def add_secure_port(self, address, server_credentials):
+
+    Args:
+      shutdown_handler:  A function taking a single arg, a time in seconds that
+      the handler should complete within.  None indicates that that handler can
```
Nit: "within which the handler should complete".

Okay, okay: "pico-nit".

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_server.py:
```diff
@@ -672,41 +677,44 @@ def _serve(state):
+  shutdown_event = threading.Event()
+
+  def cancel_all_calls_after_grace():
+    do_shutdown = False
```
Please favor
```
if <condition>:
  <name> = <expression>
else:
  <name> = <other expression>
```
over
```
<name> = <other expression>
if <condition>:
  <name> = <expression>
```
. Here, the general form particularizes to
```
<*no* assignment of do_shutdown>
with state.lock:
  if <one condition>:
    <statements that return>
  elif <another condition>:
    <statements that don't return>
    do_shutdown = True
  else:
    do_shutdown = False
```
.

kpayson64 commented:
Addressed comments.  The untested code path is almost impossible to test.

The join() function that exposes it is only called by the internals of threading.py, and we don't hold a reference to the server thread to explicitly join() it, and I'm not sure we should add a reference just for this purpose.

Moreover, the threading.py only calls join(None), the code path in question isn't even exercised by the internals of threading.py.  Its just there to make sure we properly follow the join(timeout) semantics.

nathanielmanistaatgoogle commented:
Change content looks good, but the detailed response you just gave me about the difficulty of testing some parts of this change belongs in the commit message.

kpayson64 commented:
Addressed comment.

nathanielmanistaatgoogle commented:
In your commit log message I think you mean "It's just there" rather than "Its just there".

kpayson64 commented:
@soltanmm-google ping

soltanmm-google commented on src/python/grpcio/grpc/_server.py:
```diff
@@ -745,11 +758,14 @@ def add_insecure_port(self, address):
 
-  def stop(self, grace):
-    return _stop(self._state, grace)
+  def stop(self, grace, shutdown_handler_grace=None):
```
nit-question: any particular reason why this is called `shutdown_handler_grace` and other places call it `exit_handler_grace`?

kpayson64 replied:
```shutdown_handler_grace``` and ```exit_handler_grace``` are different things.  ```shutdown_handler_grace``` is the grace period passed to to the shutdown handlers, ```exit_handler_grace``` is the ```shutdown_handler_grace``` to be used in the event of a server running on exit.  I've renamed ```exit_handler_grace``` ->```exit_shutdown_handler_grace``` to make that more clear.

kpayson64 commented:
Addressed comments, back to @soltanmm-google

soltanmm-google commented:
lgtm

nathanielmanistaatgoogle commented:
(Squash before merging, please.)
